# 236523 - Introduction to Bioinformatics 
## ![image](https://user-images.githubusercontent.com/43007010/129244877-92615d82-ad39-429d-ad8d-f5eb8c5be091.png)

Ben Filiarsky	, Yotam Martin

## Abstract
Schizophrenia (SZ) and Bipolar Disorder (BD) are serious neuropsychiatric disorders, severely harming patients' life quality. In this work, we perform a GWAS on RNAseq reads obtained from postmortem brain samples, from three different brain regions. We perform differential expression analysis (DEA) and successfully recover results of the paper our work is based on (Hu et al., 2016). As a part of our analysis, we compare DEA results with no normalization by housekeeping genes (HKG), with normalization by a general HKG set and with a gene set specifically tailored for postmortem brain samples. We show that using a custom-fit gene set for such correction is beneficial and encourage future usage of specifically tailored gene sets for correction. Significantly up and down regulated genes are then enriched using KEGG enrichment, resulting in a substantial enrichment of the MAPK signaling pathway in SZ samples in the BA9 brain region, a pathway that had been previously linked to SZ. An additional weighted gene co-expression analysis (WGCNA) combined with clustering was performed, followed by functional enrichment (GO and DO) of each cluster. In all the experiments we found a cluster with a significantly enriched term that had been tightly connected to SZ and BD development in previous works. We finish our work with raising ethical concerns regarding usage of such research in future applications. Our code repository and all the data for this work as well as results and plots are available via GitHub.

#### Project tree:
```
.
â”œâ”€â”€ Differential Expression Analysis                              | DEA results
â”‚Â Â  â”œâ”€â”€ Differential Expression results brain norm                | DEA results with brain HKG normalization
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ exp1 brain schizophrenia VS. normal BA9 .csv          | 
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ...
â”‚Â Â  â”‚Â Â  â””â”€â”€ exp6 brain bipolar_disorder VS. normal BA24 .csv
â”‚Â Â  â”œâ”€â”€ Differential Expression results general norm              | DEA results with general HKG normailzation
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ exp1 general schizophrenia VS. normal BA9 .csv
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ...
â”‚Â Â  â”‚Â Â  â””â”€â”€ exp6 general bipolar_disorder VS. normal BA24 .csv
â”‚Â Â  â”œâ”€â”€ Differential Expression results no norm                   | Initial results with no HKG correction
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ exp1 no_norm schizophrenia VS. normal BA9 .csv
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ...
â”‚Â Â  â”‚Â Â  â””â”€â”€ exp6 no_norm bipolar_disorder VS. normal BA24 .csv
â”‚Â Â  â””â”€â”€ DifferentialExpressionPlots                               | ğŸ“ŠğŸ“Š Plots for DEA
â”‚Â Â      â”œâ”€â”€ Heatmap_down                                          | pheatmap of the down regulated significant genes
â”‚Â Â      â”‚Â Â  â”œâ”€â”€ exp1
â”‚Â Â      â”‚Â Â  â”‚Â Â  â”œâ”€â”€ brain.tiff
â”‚Â Â      â”‚Â Â  â”‚Â Â  â”œâ”€â”€ general.tiff
â”‚Â Â      â”‚Â Â  â”‚Â Â  â””â”€â”€ no_norm.tiff
â”‚Â Â      â”‚Â Â  â””â”€â”€ ...
â”‚Â Â      â”œâ”€â”€ Heatmap_up                                            | pheatmap of the up regulated significant genes
â”‚Â Â      â”œâ”€â”€ Kmeans_tSNE                                           | Kmeans (k=2) results 
â”‚Â Â      â”œâ”€â”€ QQ                                                    | QQ-plots of the adjusted pvalues
â”‚Â Â      â””â”€â”€ Volcano                                               | Volcano plots with annotation of significant genes
â”œâ”€â”€ Differential Expression Analysis.Rmd                          | ğŸ’»ğŸ’» Rmd Code to reproduce DEA
â”œâ”€â”€ E-GEOD-78936-atlasExperimentSummary.Rdata                     | ğŸ’¿ğŸ’¿ Original data
â”œâ”€â”€ E-GEOD-78936-query-results.tsv                                | ğŸ’¿ğŸ’¿ Original data
â”œâ”€â”€ E-GEOD-78936-raw-counts.tsv                                   | ğŸ’¿ğŸ’¿ Original data
â”œâ”€â”€ Pathway Enrichment Analysis                                   | Results of pathfindR enrichment results
â”‚Â Â  â””â”€â”€ pathfindR
â”‚Â Â      â”œâ”€â”€ brain_norm
â”‚Â Â      â”‚Â Â  â””â”€â”€ pathfindR_Results_exp_1_brain
â”‚Â Â      â”œâ”€â”€ general_norm
â”‚Â Â      â”‚Â Â  â””â”€â”€ pathfindR_Results_exp_1_general
â”‚Â Â      â””â”€â”€ no_norm
â”‚Â Â          â”œâ”€â”€ pathfindR_Results_exp_1
â”‚Â Â          â”œâ”€â”€ pathfindR_Results_exp_1_3_5
â”‚Â Â          â”œâ”€â”€ pathfindR_Results_exp_3
â”‚Â Â          â””â”€â”€ pathfindR_Results_exp_5
â”œâ”€â”€ Pathway Enrichment Analysis.Rmd                               | ğŸ’»ğŸ’» Rmd Code to reproduce enrichment analysis
â”œâ”€â”€ Weighted Gene Co-Expression Analysis                          | ğŸŒğŸŒ Results for WGCNA 
â”‚Â Â  â”œâ”€â”€ Nets
â”‚Â Â  â””â”€â”€ Plots                                                     | ğŸ“ŠğŸ“Š Plots for GO/DO enrichment
â”‚Â Â      â”œâ”€â”€ DO
â”‚Â Â      â”œâ”€â”€ GO
â”œâ”€â”€ Weighted Gene Co-Expression Analysis.Rmd                      | ğŸ’»ğŸ’» Rmd Code to reproduce WGCNA
â”œâ”€â”€ compare_results_to_paper.py                                   | ğŸ Comparing results to Hu et al. (2016) 
â”œâ”€â”€ find_lncRNA.py                                                | ğŸ Arrange found genes table in DEA
â”œâ”€â”€ genes_transcript_type_ncbi_kegg.csv                           | ğŸ Arrange found genes table in DEA
â”œâ”€â”€ genes_transcript_type_ncbi_kegg_unique.csv                    | ğŸ Arrange found genes table in DEA
â”œâ”€â”€ housekeeping_brain.csv                                        | Brain HKG list
â”œâ”€â”€ housekeeping_general.csv                                      | General HKG list
â”œâ”€â”€ general and brain keeping genes processing.py                 | ğŸ Organize HKG lists
â”œâ”€â”€ significant_genes_by_experiment_brain.csv                     | Detailed table of significant genes (Exp1-6, brain HKG normalization)
â”œâ”€â”€ significant_genes_by_experiment_general.csv                   | Detailed table of significant genes (Exp1-6, general HKG normalization)
â”œâ”€â”€ significant_genes_by_experiment_no_norm.csv                   | Detailed table of significant genes (Exp1-6, no HKG normalization)
â”œâ”€â”€ significant_genes_exp_1_to_6.py                               | ğŸ Find the significant genes from Exp1-6
â”œâ”€â”€ up_down_reg_genes_per_experiment.csv                          | Table of significant genes (Exp1-6, no HKG normalization)
â”œâ”€â”€ up_down_reg_genes_per_experiment_brain.csv                    | Table of significant genes (Exp1-6, brain HKG normalization)
â””â”€â”€ up_down_reg_genes_per_experiment_general.csv                  | Table of significant genes (Exp1-6, general HKG normalization)
```
